Plane: add crow flap options parameter, allows 'full house' wing

IamPete1 opened the pull request:
fix for https://github.com/ArduPilot/ardupilot/issues/8009 and https://github.com/ArduPilot/ardupilot/issues/10036

Allow a 'full house' glider as outlined here https://www.youtube.com/watch?v=gqjIg6fB_fQ only missing camber control but I think that is probably not that useful in our case. 

This adds a new bit mask and a parameter. The bit mask allows three features to be enabled or disabled. 

- use aileron vs elevon as input for mixer, current code uses elevon so can only be used for flying wings
- full span ailerons, current code does this, now it can be turned off so only the outer flaps are used for aileron and inner used for differential rudder and flaps
- progressive flap, current code puts outer up as inner go down, progressive flap puts inner down first 0 to 50% flap, then from 50 to 100% the outers are put up. I think this is more useful, half can then be used for just flaps for extra lift and full gives full crow for stopping. 

This also adds a parameter to allow flap down travel to be limited to a percentage of full travel when used as aileron and in differential rudder. This allows flaps with lots of down travel to be used as full span ailerons.

By default the parameters give no change to current behavior, 

vid of SITL testing: https://youtu.be/25VIhl6fEPw

IamPete1 commented:
Set this up on my glider this evening, all works as it should

IamPete1 commented:
rebased and flight tested

samuelctabor commented:
It would be nice to separate the two functions. Flap gets more lift, airbrake additional drag, and crow mixes the two. There is already the functionality for flap, and the progressive flap here makes this do airbrake as well, but allowing a separate mechanism to specify airbrake would allow for use of a separate input channel, or control glideslope during auto landing. The user could do a TX mix to get the progressive flap functionality.

IamPete1 commented:
I did think about linking into the reverse thrust stuff but decided that the autopilot wouldn't know what to do with brakes

samuelctabor commented:
I have tested autolanding using airbrakes and the reverse thrust feature. I haven't pushed it yet but it does work very well. It would be nice for crow to be able to use it too

IamPete1 commented:
We can add another option to the bitmask to do exactly that when that goes in. Do you have a another rc_input set for airbrakes too? or just leave it all to the auto stuff? For the flaps this takes the max of the the manual and auto flaps, that seems like a sensible thing to do for air-brakes also

samuelctabor commented:
This is my TECS PR for gliders mentioned on the call https://github.com/ArduPilot/ardupilot/pull/9643

samuelctabor commented:
I do, yes. It's in this branch https://github.com/samuelctabor/ardupilot/commits/autoAirbrakes_394

I need to clean it up a bit, currently I think it breaks actual reverse thrust.

IamPete1 commented:
@tridge I have updated the parameter distinctions as discuss in the devcall and added a glider model for testing here https://github.com/ArduPilot/SITL_Models/pull/6 Thanks

tridge commented:
@IamPete1 I went to test this with the Mistral3m and realised I also need the RFX file with expo removed, mixers removed and low rates removed. Can you add that to SITL_Models?

IamPete1 commented:
rebased on master

IamPete1 commented:
seems to have failed Travis due to copter auto test, must be unrelated to this PR

kd0aij commented:
I restarted the failed test. It's been failing a lot lately for no apparent reason.

kd0aij commented:
@IamPete1 @tridge bungee launch and chase view are available in RF8 with a 3D airport

https://youtu.be/ekoagXKBN-k

IamPete1 commented:
@tridge  mid-week reminder ;), I have just re based

magicrub commented:
I looked this over and it seems OK to me. No testing, just a code review. @tridge has really wanted to take a deep dive and I hope he does. Looks like a great addition!

tridge commented:
nice work, sorry it took me so long to review!
we really need a wiki page on this now

IamPete1 commented:
Thanks, i will update the wiki now its in
